### PR TITLE
made first successful io network connection :D

### PR DESCRIPTION
**Pull request by lancejpollard**
@anthonyshort @stevenmiller888 This is the basics of how the dataflow stuff would work, does it make sense?

Bare in mind this is super raw and ugly, needs to be cleaned up a lottttt.

Holler if you want me to explain any of it.

### CODE REVIEW

**Comment by lancejpollard** [comment]
basic example:

```js
var Test = component(template)
  .prop('text', value('meta').pick('title'));

var app = deku()
  .set('renderImmediate', true)
  .source('meta', metaSource);

app.mount(document.body, Test);
assert.equal(document.body.innerHTML, '<div>Hello World</div>');

function template(props) {
  return dom('div', {}, props.text);
}

function metaSource(app) {
  app.send('update value', {
    type: 'meta',
    data: { title: 'Hello World' }
  });
}
```

**Comment by anthonyshort** [comment]
This is sweet, it'll finally solve the problem that no one else has solved yet haha 

Just going to need to get my head around how it's wired up. I think the API could be a bit simpler. Random idea:

```
{
props: 
  text: {
    source: 'meta', 
    transducer: function(value){
      return value.pick('title')
    }
  }
}
```

Just so the `Component` object is just a plain, declarative object explaining the data. Even that example relies on knowledge of what the `value` object looks like.

Using the `value()` function seems a little weird. We could use it internally, but for an external API it could be confusing. The api for sources is a little confusing to me too, using `app.send` with an event name. I think it would be less confusing if it was just a method like `app.source('meta').update({ title: 'foo' })` or something. I just find events in general hard to follow.

**Comment by lancejpollard** [comment]
@anthonyshort awesome. Yeah that `value` object/dsl is just a hack for now. Ideally it would be like your saying, just a plain JSON object with perhaps a function to do the transform. Just need to think about that more.

**Comment by joshrtay** [comment]
Is the goal here to have relay like behavior without being tied to graphQL?

**Comment by anthonyshort** [comment]
A little bit. But it won't allow querying at the component level. With GraphQL each component can specify exactly what fields they want from a giant, global store and that will get inserted into the component as props. 

We're just looking to make streams of data available to components. These streams could be collections of models (like projects, users, events etc) or even a router streaming changes. You can filter these streams and the renderer will wire it up in the most optimized way. It's not as advanced as GraphQL but I think it will cover most use cases while remaining pretty simple.

@lancejpollard can probably explain this better than I can :)

**Comment by lancejpollard** [comment]
@anthonyshort Other potential thoughts from me and @stevenmiller888 messing around

```js

var Page = component()
  .prop('title', { source: 'pages', transforms: title })
  .prop('description', description)

function title(events) {
  return {
    transforms: [
      { filter: 'active' },
      { select: 'first' },
      { pick: 'name' }
    ]
  };
}

function description() {
  return {
    source: 'events',
    transforms: [
      { filter: 'active' },
      { select: 'first' },
      { pick: 'description' }
    ]
  };
}
```

**Comment by joshrtay** [comment]
@anthonyshort sounds really cool.

**Comment by stevenmiller888** [comment]
@lancejpollard @anthonyshort I think I like the first one more:

```
var Page = component()
  .prop('title', { source: 'pages', transforms: title });

function title(events) {
  return {
    transforms: [
      { filter: 'active' },
      { select: 'first' },
      { pick: 'name' }
    ]
  };
}
```
It's nice because now the component definition contains the complete picture of all external information (props + sources).

Could potentially even pull the transforms out of the definition:

```
var Page = component()
  .prop('title', { source: 'pages' });
```

and map the `title` prop to a transformation function named `title` that lives outside the component.

**Comment by anthonyshort** [comment]
:+1:  I kind of like the idea of the transforms being done elsewhere so the component doesn't need to know about the implementation. It's just getting some prop from some source, then we can define how that source is create at the app level.

**Comment by joshrtay** [comment]
@anthonyshort have you guys seen [Netflix's falcorjs](https://www.youtube.com/watch?t=3367&v=z8UgDZ4rXBU)? How would you envision deku working with a system like that? Can you implement async props or async transforms in deku?

**Comment by anthonyshort** [comment]
I'll have to check that out later tonight. If it's just a big store of all your data, it could work really nice. 

I think in general we want to be really agnostic when it comes to side-loading data into the tree. At the moment you can just set arbitrary data on the tree and then your components can just access it:

```
app.set('myEvents', [{ name: 'foo' }])
```
```
var List = {
  props: {
    'events': { source: 'myEvents' }
  }
}
```

It'll just update every component that needs that source whenever it changes. This keeps the implementation out the component and should hopefully cover most use cases. All the async stuff can happen outside of the tree and then you just set it when you're ready.
